nifi-extension-bundles/nifi-poi-bundle/nifi-poi-services/src/main/java/org/apache/nifi/excel/ExcelHeaderSchemaStrategy.java

```diff
@@ -0,0 +1,158 @@
+        }
+
+        final String requiredSheetsDelimited = context.getProperty(ExcelReader.REQUIRED_SHEETS).evaluateAttributeExpressions(variables).getValue();
+        final List<String> requiredSheets = ExcelReader.getRequiredSheets(requiredSheetsDelimited);
+        final Integer rawFirstRow = context.getProperty(ExcelReader.STARTING_ROW).evaluateAttributeExpressions(variables).asInteger();
+        final int firstRow = rawFirstRow == null ? NumberUtils.toInt(ExcelReader.STARTING_ROW.getDefaultValue()) : rawFirstRow;
+        final int zeroBasedFirstRow = ExcelReader.getZeroBasedIndex(firstRow);
+        final String password = context.getProperty(ExcelReader.PASSWORD).getValue();
+        final ExcelRecordReaderConfiguration configuration = new ExcelRecordReaderConfiguration.Builder()
+                .withRequiredSheets(requiredSheets)
+                .withFirstRow(zeroBasedFirstRow)
+                .withPassword(password)
+                .build();
+
+        final RowIterator rowIterator = new RowIterator(contentStream, configuration, logger);
+        final Map<String, FieldTypeInference> typeMap = new LinkedHashMap<>();
+        List<String> fieldNames = null;
+        int index = 0;
+
+        while (rowIterator.hasNext()) {
+           Row row = rowIterator.next();
+           if (index == 0) {
+               fieldNames = getFieldNames(firstRow, row);
+           } else if (index <= NUM_ROWS_TO_DETERMINE_TYPES) {
+               inferSchema(row, fieldNames, typeMap);
+           } else {
+               break;
+           }
+
+           index++;
+        }
+
+        return createSchema(typeMap);
+    }
+
+    private List<String> getFieldNames(int firstRowIndex, Row row) throws SchemaNotFoundException {
+        if (!ExcelUtils.hasCells(row)) {
+            throw new SchemaNotFoundException(String.format("Field names could not be determined from configured header row %s, as this row has no cells with data", firstRowIndex));
+        }
+
+        final List<String> fieldNames = new ArrayList<>();
+        for (int index = 0; index < row.getLastCellNum(); index++) {
+            final Cell cell = row.getCell(index);
+            final String fieldName = dataFormatter.formatCellValue(cell);
+
+            if (fieldName == null || fieldName.isEmpty()) {
+                throw new SchemaNotFoundException(String.format("Configured header row %s at cell %s is empty which is not suitable for a field name", row.getRowNum(), index));
+            }
+
+            fieldNames.add(fieldName);
+        }
+
+        return fieldNames;
+    }
+
+    private void inferSchema(final Row row, final List<String> fieldNames, final Map<String, FieldTypeInference> typeMap) throws SchemaNotFoundException {
+        if (!ExcelUtils.hasCells(row)) {
+            throw new SchemaNotFoundException(String.format("Row %s a required row for inferring field types has no cells", row.getRowNum()));
```

Comment:
This check effectively requires a minimum of 10 rows, which should not be necessary.

Instead of requiring 10 rows and requiring a matching number of cells, it seems better to make this method more permissive. That would allow less than 10 rows, and it would allow some empty or partial rows, but that seems acceptable because the type inference is a best-effort approach.

Reply:
Per the logic on lines 91-99 this header strategy can work with less than ten lines if the data is less than 10 lines long. Furthermore my logic allows for empty cells but the total number of cells cannot exceed the initial size of the header (lines 132-134) which I thought makes sense as I need to align each field to a known column. If there all of a sudden is a new column that would throw things off.

Reply:
Thanks for the reply. Yes, I think encapsulating the logic to ignore empty rows would accomplish the end goal.